lib/stdlib/src/gen_statem.erl

```diff
@@ -845,109 +850,117 @@ loop_hibernate(Parent, Debug, S) ->
        {wakeup_from_hibernate,3}}).
 
 %% Entry point for wakeup_from_hibernate/3
-loop_receive(Parent, Debug, S) ->
+loop_receive(Parent, Debug, #{cancel_timers := 0, auto_hibernate_timeout := AutoHibernateTimeout} = S) when is_integer(AutoHibernateTimeout) ->
```

Comment:
The value `'infinity'` is allowed as an `after` argument - all values of type `timeout()` are, so there is no need to test `is_integer(AutoHibernateTimeout)`.  Just use the value in `after` without testing.  If there should be a test of the value it should have been done when obtaining the option value (at the API).

As I see it you should not need to test for `cancel_timers := 0` since that test is used to delay hibernation ordered by the state machine implementation when that coincides with cancelling timer(s).  If the state machine engine has to wait for a timer cancel message for longer than `AutoHibernateTimeout`, so be it; very strange, might as well hibernate.  Unless there is something I missed, of course.

So you should be able get away with just augmenting the `receive` clause in `loop_recevie/3` with an `after AutoHibernateTimeout` clause and not need to add the `handle_received_msg/3` function.